pkg/object/mqttproxy/session.go

```diff
@@ -0,0 +1,246 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package mqttproxy
+
+import (
+	"encoding/base64"
+	"sync"
+	"time"
+
+	"gopkg.in/yaml.v2"
+
+	"github.com/eclipse/paho.mqtt.golang/packets"
+	"github.com/megaease/easegress/pkg/logger"
+)
+
+type (
+	// SessionInfo is info about session that will be put into etcd for persistency
+	SessionInfo struct {
+		// map subscribe topic to qos
+		EGName    string         `yaml:"egName"`
+		Name      string         `yaml:"name"`
+		Topics    map[string]int `yaml:"topics"`
+		ClientID  string         `yaml:"clientID"`
+		CleanFlag bool           `yaml:"cleanFlag"`
+	}
+
+	// Session includes the information about the connect between client and broker,
+	// such as topic subscribe, not-send messages, etc.
+	Session struct {
+		sync.Mutex
+		broker       *Broker
+		storeCh      chan SessionStore
+		info         *SessionInfo
+		done         chan struct{}
+		pending      map[uint16]*Message
+		pendingQueue []uint16
+		nextID       uint16
+	}
+
+	// Message is the message send from broker to client
+	Message struct {
+		Topic      string `yaml:"topic"`
+		B64Payload string `yaml:"b64Payload"`
+		Qos        int    `yaml:"qos"`
+	}
+)
+
+func newMsg(topic string, payload []byte, qos byte) *Message {
+	m := &Message{
+		Topic:      topic,
+		B64Payload: base64.StdEncoding.EncodeToString(payload),
+		Qos:        int(qos),
+	}
+	return m
+}
+
+func (s *Session) store() {
+	str, err := s.encode()
+	if err != nil {
+		logger.Errorf("encode session %+v failed, %v", s, err)
+		return
+	}
+	ss := SessionStore{
+		key:   s.info.ClientID,
+		value: str,
+	}
+	go func() {
+		s.storeCh <- ss
+	}()
+}
+
+func (s *Session) encode() (string, error) {
+	b, err := yaml.Marshal(s.info)
+	if err != nil {
+		return "", err
+	}
+	return string(b), nil
+}
+
+func (s *Session) decode(str string) error {
+	return yaml.Unmarshal([]byte(str), s.info)
+}
+
+func (s *Session) init(sm *SessionManager, b *Broker, connect *packets.ConnectPacket) error {
+	s.broker = b
+	s.storeCh = sm.storeCh
+	s.done = make(chan struct{})
+	s.pending = make(map[uint16]*Message)
+	s.pendingQueue = []uint16{}
+
+	s.info = &SessionInfo{}
+	s.info.EGName = b.egName
+	s.info.Name = b.name
+	s.info.ClientID = connect.ClientIdentifier
+	s.info.CleanFlag = connect.CleanSession
+	s.info.Topics = make(map[string]int)
+	return nil
+}
+
+func (s *Session) updateEGName(egName, name string) {
+	s.Lock()
+	defer s.Unlock()
+	s.info.EGName = egName
+	s.info.Name = name
+	s.store()
+}
+
+func (s *Session) subscribe(topics []string, qoss []byte) error {
+	s.Lock()
+	defer s.Unlock()
+	for i, t := range topics {
+		s.info.Topics[t] = int(qoss[i])
+	}
+	s.store()
+	return nil
+}
+
+func (s *Session) unsubscribe(topics []string) error {
+	s.Lock()
+	defer s.Unlock()
+	for _, t := range topics {
+		delete(s.info.Topics, t)
+	}
+	s.store()
+	return nil
+}
+
+func (s *Session) allSubscribes() ([]string, []byte, error) {
```

Comment:
I think we don't need an `error` among this function's return values. Cause it's always `nil`.

Reply:
Here I make `Session.allSubscribes` and other methods in `Session` as general apis, return error will make it easy for further update. Although now, we don't need to return error, but if one day we change our design we only need to change the `Session` method implement and not need to change other module. I don't know if this is a good approach... Looking forward for your advice about this.

Reply:
I think this is a `Class.Getter` like operation in C++/Java, it gets its own data and transforms it to another format without any outer dependence. So I prefer to remove this error return. But it's up to u, not a very strong requirement here.